src/zmalloc.c

```diff
@@ -409,8 +423,11 @@ char *zstrdup(const char *s) {
 }
 
 size_t zmalloc_used_memory(void) {
-    size_t um;
-    atomicGet(used_memory,um);
+    assert(total_active_threads < MAX_THREADS_NUM);
+    size_t um = 0;
+    for (int i = 0; i < total_active_threads; i++) {
```

Comment:
Wondering how important is it for zmalloc_used_memory to return the correct used_memory? As we loop through  to add up the memory used by each thread, the threads that had been counted may continue to allocate to free memory, so the sum might under or over account the actual memory used by all threads. 

I don't think this is an issue if the function is only used for metrics, but overMaxmemoryAfterAlloc (evict.c) is not for metrics,  we probably should ensure (and call out) that undercounting memory usage would not be an issue there

Reply:
a few quick thoughts

1. I see value in removing atomic operations
2. Since we use `size_t` here for the counters, which should be natively aligned as well, the worst we can get would be staleness as opposed to inconsistency.
3. The absolute accuracy doesn't have much value for metrics. The value will get stale immediately after the `INFO` command returns, whether it is absolutely accurate or not. As long as the value is close enough I think we are good. We can change the tests
4. Agreed that `overMaxmemoryAfterAlloc` is a higher bar than metrics. My gut feeling is that the staleness brought by the relaxed memory access should still be negligible. I doubt we could accumulate a lot of changes before the CPU flushes out its cache to the main memory. Curious to know if anyone has more definitive information.
5. The absolute accuracy requirement is directly at odds with the performance overhead incurred by the atomic operation. So one of them has to go.

Reply:
Seems we all had concern about the absolute accuracy of `used_memory` in `overMaxmemoryAfterAlloc`, IMO, this is not a problem, because in the definition of `overMaxmemoryAfterAlloc` , the comparison between `moremem` and `used_memory` is not absolute accuracy if we consider the gap between requested memory and actual allocated memory from allocator.  
``` c
int overMaxmemoryAfterAlloc(size_t moremem) {
    if (!server.maxmemory) return  0; /* No limit. */

    /* Check quickly. */
    size_t mem_used = zmalloc_used_memory();
    if (mem_used + moremem <= server.maxmemory) return 0;

    size_t overhead = freeMemoryGetNotCountedMemory();
    mem_used = (mem_used > overhead) ? mem_used - overhead : 0;
    return mem_used + moremem > server.maxmemory;
}
```

Reply:
Kindly ping @PingXie , any comments about this, do we need to continue this?

Reply:
I think there is a way to address this concern while retaining the most of the performance benefit if we keep the existing io-threading model. The idea would be to keep a mirrored per-thread counter array such that each thread periodically "flushes" its thread local counter value to the corresponding element in the global per-thread counter array using atomic operations. This would work with the current io-threading model where io-threads and main threads run in lock-steps. Every io thread would need to push its local value to this global array when it is done with its work, but only for once. Then by the time the main thread starts executing, it would have the accurate data. We can reduce the number of atomic operations to one per io-thread per epoll session. 

We would have to revisit this if we went for a different threading model (such as #22) .

That said, I am still not sure how accurate this information needs to be.  `zmalloc_used_memory` doesn't represent the RSS usage of the process anyways. I think it would be really helpful if we could somehow establish the worst possible case, empirically or even a SWAG.

Reply:
> That said, I am still not sure how accurate this information needs to be. `zmalloc_used_memory` doesn't represent the RSS usage of the process anyways.
>

IMO, there is no meaningful of the absolute accurate `zmalloc_used_memory`, to achieve the goal, the cost is very high, not only the expensive atomic operation in every `zmalloc/zfree` operation, but also the call of `je_malloc_usable_size` is very high (**From the cycles hotspot snapshot in top comment, you can find the cycles ratio of `je_malloc_usable_size` is the top 1 hotspot**). If we can accept the a little latency of `zmalloc_used_memory`,  ideally we may at least have a 10% perf gain.

Reply:
I saw you opened #467. Thanks @lipzhu.